src/wh_server_crypto.c

```diff
@@ -1963,6 +2007,348 @@ static int _HandleSha256Dma(whServerContext* server, whPacket* packet,
+                         */
+                        whKeyId keyId =
+                            WH_MAKE_KEYID(WH_KEYTYPE_CRYPTO,
+                                          server->comm->client_id, req->keyId);
+
+                        if (WH_KEYID_ISERASED(keyId)) {
+                            /* Generate a new id */
+                            ret = hsmGetUniqueId(server, &keyId);
+#ifdef DEBUG_CRYPTOCB
+                            printf("[server] %s UniqueId: keyId:%u, ret:%d\n",
+                                   __func__, keyId, ret);
+#endif
+                        }
+
+                        if (ret == 0) {
+                            ret = wh_Server_MlDsaKeyCacheImport(
+                                server, key, keyId, req->flags, req->labelSize,
+                                req->label);
+#ifdef DEBUG_CRYPTOCB
+                            printf(
+                                "[server] %s CacheImport: keyId:%u, ret:%d\n",
+                                __func__, keyId, ret);
+#endif
+                            if (ret == 0) {
+                                res->keyId   = WH_KEYID_ID(keyId);
+                                res->keySize = keySize;
+                                *size = WH_PACKET_STUB_SIZE + sizeof(res);
+                            }
+                        }
+                    }
+                }
+            }
+            wc_MlDsaKey_Free(key);
+        }
+    }
+
+    if (ret == WH_ERROR_ACCESS) {
+        res->dmaAddrStatus.badAddr = req->key;
+    }
+
+    return ret;
+#endif /* WOLFSSL_DILITHIUM_NO_MAKE_KEY */
+}
+
+static int _HandleMlDsaSignDma(whServerContext* ctx, whPacket* packet,
+                               uint16_t* out_size)
+{
+#ifdef WOLFSSL_DILITHIUM_NO_SIGN
+    (void)ctx;
+    (void)packet;
+    (void)out_size;
+    return WH_ERROR_NOHANDLER;
+#else
+    int      ret = 0;
+    MlDsaKey key[1];
+    void*    msgAddr = NULL;
+    void*    sigAddr = NULL;
+
+#if WH_DMA_IS_32BIT
+    wh_Packet_pq_mldsa_sign_Dma32_req* req = &packet->pqMldsaSignDma32Req;
+    wh_Packet_pq_mldsa_sign_Dma32_res* res = &packet->pqMldsaSignDma32Res;
+#else
+    wh_Packet_pq_mldsa_sign_Dma64_req* req = &packet->pqMldsaSignDma64Req;
+    wh_Packet_pq_mldsa_sign_Dma64_res* res = &packet->pqMldsaSignDma64Res;
+#endif
+
+    /* Transaction state */
+    whKeyId key_id;
+    int     evict = 0;
+
+    if (ctx == NULL || packet == NULL || out_size == NULL) {
+        return WH_ERROR_BADARGS;
+    }
+
+    /* Get key ID and evict flag */
+    key_id = WH_MAKE_KEYID(WH_KEYTYPE_CRYPTO, ctx->comm->client_id, req->keyId);
+    evict  = !!(req->options & WH_PACKET_PQ_MLDSA_SIGN_OPTIONS_EVICT);
+
+    /* Initialize key */
+    ret = wc_MlDsaKey_Init(key, NULL, ctx->crypto->devId);
+    if (ret != 0) {
+        return ret;
+    }
+
+    /* Export key from cache */
+    /* TODO: sanity check security level against key pulled from cache? */
```

Comment:
The Sign operation later should get the BAD_ARGS, no?  Ok to check it early, but not required.